src/Task.h

```diff
@@ -790,6 +790,26 @@ class Task {
   void xptrace(int request, remote_ptr<void> addr, void* data);
 
   /**
+   * If the given memory region is mapped into the local address, space, invoke
+   * the callback with a pointer to the local equivalent of addr and the total
+   * number of bytes available through this mapping
+   */
+  template <typename callbackT>
+  ssize_t with_local_mapping(remote_ptr<void> addr, size_t size,
+                             callbackT callback) {
+    if (vm()->has_mapping(addr)) {
+      const AddressSpace::Mapping& map = vm()->mapping_of(addr);
+      if (map.local_addr != nullptr) {
+        size_t nbytes =
+            std::min(size, static_cast<uintptr_t>(map.map.end() - addr));
+        size_t offset = addr - map.map.start();
+        return callback(static_cast<uint8_t*>(map.local_addr) + offset, nbytes);
+      }
```

Comment:
Another possibly simpler approach would be to have the helper just return a pointer to the local data if the entire requested region is locally mapped, otherwise null. I can't think of real cases where being able to return a partial local map is useful. That simplifies things to two cases (locally mapped/not locally mapped) instead of three (locally mapped/not locally mapped/partially mapped) and means we don't need a callback.

Also, this helper can live in AddressSpace and probably should.